admin/aaxinfo.1.in
.\" Manpage for aaxinfo.
.\" Contact tech@adalin.com to correct errors or typos.
.TH man 1 "19 Dec 2014" "@AAX_UTILS_MAJOR_VERSION@.@AAX_UTILS_MINOR_VERSION@.@AAX_UTILS_MICRO_VERSION@" "aaxinfo man page"
.SH NAME
aaxinfo \- Shows information on supported audio devices.
.SH SYNOPSIS
.B aaxinfo
[\fIOPTION\fR]...
.SH DESCRIPTION
.PP
Shows information on supported audio devices.
.TP
\fB\-d\fR, \fB\-\-device \fRDEVICE\fR
show mixer and hardware support for the specified device.
.TP
\fB\-c\fR, \fB\-\-copyright
shows copyright information for AeonWave.
.SH AUTHOR
Written by Erik Hofman <tech@adalin.com>
.SH SEE ALSO
aaxplay(1), aaxcvt(1)



ChangeLog
* Fri Dec 12 2014 - tech@adalin.org
- AeonWave (2.5.0) stable; urgency=low
 * Added a capabilities interface to detect the limits of the audio hardware
   by making calls to the aaxMixerGetSetup function.
 * Changed the threading model from one thread per mixer-registered audio-frame
   to a load-balancing threadpool model. This slightly increases the work the
   CPU has to do but devides the wall time spent on rendering by the number of
   physical cores. Threading is now always active on multi-core systems except
   when the environment variable AAX_USE_THREADPOOL is set to 0.
 * Add support for the ARM architecture including support for hardware FPU
   and NEON. Rendering performance is comparible with Intel/AMD Cores
   at a matching CPU frequency.
 * Support batched (fast, non-realtime) mode for Audio Files and Loopback.
 * Support defining a default output device in the confguration file.
 * Fix a problem when a registered sensor had a different number of
   audio periods in relation to the playback device.
 * Fix an emitter looping bug introduced in version 2.4.1

 -- Erik Hofman <tech@adalin.org>  Fri, 12 Dec 2014 14:16:00 + 0100

* Tue Jun 10 2014 - tech@adalin.org
- AeonWave (2.4.1) stable; urgency=low
 * Fixed a possible segmentation fault with certain ALSA drivers.
 * Fixed a possible segmentation fault for looping emitters.

 -- Erik Hofman <tech@adalin.org>  Tue, 10 Jun 2014 12:58:00 + 0200

* Wed May 21 2014 - tech@adalin.org
- AeonWave (2.4.0) stable; urgency=low
 * Changed the conditions for binary distribution;
   The library may now be included in commercial and non commercial projects
   provided the user is made aware of the option to upgrade to the HD version.
   See Copyright section 2.1; Rights and Limitations of Grant.
 * Reorganized the code to support different hardware rendering schemes in
   the future. As a result the current rendering speed has increased 25% to 75%
   (depending of the computer hardware, older hardware profit the most).
 * Add the option to use timer-based rendering instead of interrupt-based
   rendering for ALSA (experimental).
 * Fix a problem where the WASAPI device period could become lower than the 
   minimum allowed period for that particular device.
 * Switched to a new xml configuration file format.
   Both the old and the new formats will be supported up to version 3.0.0
 * Allow custom pcm devices (defined in .asoundrc) for ALSA.
 * Tweaked Surround Sound, Spatial and HRTF playback.
 * Support WAV files with a LIST section.
 * Fix Auto generated waveform loop-points.
 * Fix a problem with capture-only backends

 -- Erik Hofman <tech@adalin.org>  Wed, 21 Mar 2014 10:45:00 +0200

* Sat Dec 21 2013 - tech@adalin.org
- AeonWave (2.3.4) stable; urgency=high
 * Fix a problem where distance delayed objects that where registered at
   the mixer or an audio-frame in non-playing mode started with the wrong 
   position and orientation matrix causing unwanted delays.
 * Fix an incorrect header problem while streaming mp3 files.
 * Fix a stupid error where AeonWave-Lite didn't initialize correctly.
 * Fix volume handling for windows devices without hardware volume support.
 * Activate AVX instruction support.

 -- Erik Hofman <tech@adalin.org>  Sat, 21 Dec 2013 11:21:00 +0100

* Mon Nov 25 2013 - tech@adalin.org
- AeonWave (2.3.3) stable; urgency=medium
 * Fix a problem with files with a .wav extension that aren't.
 * Prevent destroying of the emitter unless it is in the AAX_PROCESSED state.

 -- Erik Hofman <tech@adalin.org>  Mon, 25 Nov 2013 12:34:00 +0100

* Tue Nov 21 2013 - tech@adalin.org
- AeonWave (2.3.2) stable; urgency=medium
 * Fix clipping during spoken text under certain conditions.
 * Fix a problem where the equalizer was turned off under certain conditions.
 * Fix incorrect file-type handling for the Audio Files backend and add proper
   error reporting to the calling program.

 -- Erik Hofman <tech@adalin.org>  Tue, 21 Nov 2013 13:21:00 +0100

* Sun Nov 14 2013 - tech@adalin.org
- AeonWave (2.3.1) stable; urgency=high
 * Complete rewrite of the File backend
  - Add mp3 file output support if libmp3lame is found.
  - Add support for more WAV file types and fix mulaw and alaw decoding.
 * Add the option to register one sensor opened in write mode (file only) at
   the mixer to support a dedicated record to file option.
 * Fix a major registered-sensor bug (most noticeable on Windows).
 * Fix an occasional deadlock when deregistering audio-frames.

 -- Erik Hofman <tech@adalin.org>  Sun, 14 Nov 2013 15:14:00 +0200

* Fri Oct 18 2013 - tech@adalin.org
- AeonWave (2.3.0) stable; urgency=high
 * Distance-delay mode now adds delayed state changes in addition to delayed
   start/stop. State changes include; pitch, gain, position, orientation and
   velocity.
 * Added sense of height to Surround sound; emitters can now be heard above
   and below the speakers. This is best perceived if the speakers and the
   listeners ears are (almost) in the same horizontal plane.
 * Add support for the AAXS (AeonWave Auto-generated XML Sound) audio format.
 * Add double precision 64-bit floating point matrix support functions.
 * Implement hardware volume support for ALSA and WASAPI.
   - hardware volume is adjusted within hardware boundaries, volume settings
     outside the hardware range are adjusted using software fall-back.
   - the volume filter now reports the hardware minimum and maximum.
 * Implement Auto-Gain Control with a user definable adjust-rate.
 * Add the option to select stereo, mono-left, mono-right or mono-mix for
   audio capturing.
 * Add an aaxMixerGetState function to retrieve the state of the current
   audio stream. This can for instance be used to reinitialize the device
   after the hardware audio format has changed.
 * Add the option to set the number of times to loop a buffer.
 * Fixed reported bugs.
   - Fix an issue that prevented ALSA from reporting new hardware while an
     application was running.

src/aaxinfo.c
/*
 * Copyright (C) 2008-2014 by Erik Hofman.
 * Copyright (C) 2009-2014 by Adalin B.V.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 *    1. Redistributions of source code must retain the above copyright notice,
 *        this list of conditions and the following disclaimer.
 * 
 *    2. Redistributions in binary form must reproduce the above copyright
 *        notice, this list of conditions and the following disclaimer in the
 *        documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY ADALIN B.V. ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN
 * NO EVENT SHALL ADALIN B.V. OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES 
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR 
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUTOF THE USE 
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * The views and conclusions contained in the software and documentation are
 * those of the authors and should not be interpreted as representing official
 * policies, either expressed or implied, of Adalin B.V.
 */

#if HAVE_CONFIG_H
# include "config.h"
#endif

#ifdef HAVE_LIMITS_H
# include <limits.h>
#endif

#include <stdio.h>
#include <string.h>
#if HAVE_UNISTD_H
# include <unistd.h>
#endif
#if HAVE_SYS_IOCTL_H
# include <sys/ioctl.h>
#endif
#if _WIN32
# include <Windows.h>
#endif
#ifndef NDEBUG
# ifdef HAVE_RMALLOC_H
#  include <rmalloc.h>
# endif
#endif
#include <aax/aax.h>
#include "wavfile.h"
#include "driver.h"

static const char *_filter_s[AAX_FILTER_MAX] =
{
    "_",
    "AAX_equalizer",		/* 2d filters */
    "AAX_graphic_equalizer",
    "AAX_compressor",
    "AAX_volume_filter",
    "AAX_dynamic_gain_filter",
    "AAX_timed_gain_filter",
    "AAX_frequency_filter",

    "AAX_angular_filter",
    "AAX_distance_filter"	/* 3d filters */
};

static const char *_effect_s[AAX_EFFECT_MAX] =
{
    "_",
    "AAX_pitch_effect",		/* 2d effects */
    "AAX_dynamic_pitch_effect",
    "AAX_timed_pitch_effect",
    "AAX_phasing_effect",
    "AAX_chorus_effect",
    "AAX_flanging_effect",
    "AAX_distortion_effect",
    "AAX_reverb_effect",

    "AAX_velocity_effect"	/* 3d effects */
};

static int maximumWidth = 80;

#if _WIN32
static int
terminalWidth()
{
    CONSOLE_SCREEN_BUFFER_INFO csbi;
    int ret, rv = maximumWidth;
    ret = GetConsoleScreenBufferInfo(GetStdHandle( STD_OUTPUT_HANDLE ),&csbi);
    if (ret) rv = csbi.dwSize.X;
    return rv;
}
#elif HAVE_SYS_IOCTL_H
static int
terminalWidth()
{
    struct winsize w;
    ioctl(STDOUT_FILENO, TIOCGWINSZ, &w);
    return w.ws_col;
}
#else
#pragma warning Implement terminal width
static int
terminalWidth() {
    return maximumWidth;
}
#endif

int main(int argc, char **argv)
{
    unsigned int i, x, y, z, max;
    aaxConfig cfg;
    const char *s;
    char *devname;
    int mode;

    if (printCopyright(argc, argv) || playAudioTune(argc, argv)) {
        return 0;
    }

    maximumWidth = terminalWidth()-1;
    printf("aaxinfo version %i.%i.%i\n", AAX_UTILS_MAJOR_VERSION,
                                         AAX_UTILS_MINOR_VERSION,
                                         AAX_UTILS_MICRO_VERSION);
    printf("Run %s -copyright to read the copyright information.\n", argv[0]);

    for (mode = AAX_MODE_READ; mode <= AAX_MODE_WRITE_STEREO; mode++)
    {
        char *desc[2] = { "capture", "playback"};

        printf("\nDevices that support %s:\n", desc[mode]);

        max = aaxDriverGetCount(mode);
        for (x=0; x<max; x++)
        {
            cfg = aaxDriverGetByPos(x, mode);
            if (cfg) {
                unsigned max_device;
                const char *d;

                d = aaxDriverGetSetup(cfg, AAX_DRIVER_STRING);
                max_device = aaxDriverGetDeviceCount(cfg, mode);
                if (max_device)
                {
                    for (y=0; y<max_device; y++)
                    {
                        unsigned int max_interface;
                        const char *r;

                        r = aaxDriverGetDeviceNameByPos(cfg, y, mode);

                        max_interface= aaxDriverGetInterfaceCount(cfg, r, mode);
                        if (max_interface)
                        {
                            for (z=0; z<max_interface; z++)
                            {
                                const char *ifs;

                                ifs = aaxDriverGetInterfaceNameByPos(cfg, r, z,
                                                                     mode);
                                printf(" '%s on %s: %s'\n", d, r, ifs);
                            }
                        }
                        else {
                           printf(" '%s on %s'\n", d, r);
                        }
                    }
                }
                else {
                    printf(" '%s'\n", d);
                }
                aaxDriverClose(cfg);
                aaxDriverDestroy(cfg);
            } else {
                printf("\t%i. not found\n", x);
            }
        }
    }

    mode = AAX_MODE_WRITE_STEREO;
    devname = getDeviceName(argc, argv);
    cfg = aaxDriverGetByName(devname, mode);
    if (cfg)
    {
        printf("\n");
        cfg = aaxDriverOpen(cfg);
        if (cfg)
        {
            int res, min, max;

            res = aaxMixerSetState(cfg, AAX_INITIALIZED);
            testForState(res, "aaxMixerInit");

            s = aaxDriverGetSetup(cfg, AAX_DRIVER_STRING);
            printf("Driver  : %s\n", s);

            s = aaxDriverGetSetup(cfg, AAX_RENDERER_STRING);
            printf("Renderer: %s\n", s);

            x = aaxGetMajorVersion();
            y = aaxGetMinorVersion();
            s = (char *)aaxGetVersionString(cfg);
            printf("Version : %s (%i.%i)\n", s, x, y);

            s = aaxDriverGetSetup(cfg, AAX_VENDOR_STRING);
            printf("Vendor  : %s\n", s);

            x = aaxMixerGetSetup(cfg, AAX_TIMER_MODE);
            printf ("Mixer timed mode support:   %s\n", x ? "yes" : "no");

            x = aaxMixerGetSetup(cfg, AAX_SHARED_MODE);
            printf ("Mixer shared mode support:  %s\n", x ? "yes" : "no");

            x = aaxMixerGetSetup(cfg, AAX_BATCHED_MODE);
            printf ("Mixer batched mode support: %s\n", x ? "yes" : "no");

            min = aaxMixerGetSetup(cfg, AAX_TRACKS_MIN);
            max = aaxMixerGetSetup(cfg, AAX_TRACKS_MAX);
            printf("Mixer supported track range: %i - %i tracks\n", min, max);

            min = aaxMixerGetSetup(cfg, AAX_FREQUENCY_MIN);
            max = aaxMixerGetSetup(cfg, AAX_FREQUENCY_MAX);
            printf("Mixer frequency range: %4.1fkHz - %4.1fkHz\n",
                    min/1000.0f, max/1000.0f);

            x = aaxMixerGetSetup(cfg, AAX_FREQUENCY);
            printf("Mixer frequency: %i Hz\n", x);

            x = aaxMixerGetSetup(cfg, AAX_REFRESHRATE);
            printf("Mixer refresh rate: %u Hz\n", x);

            x = aaxMixerGetSetup(cfg, AAX_UPDATERATE);
            if (x) {
                printf("Mixer update rate:  %u Hz\n", x);
            }

            x = aaxMixerGetSetup(cfg, AAX_LATENCY);
            if (x) {
               printf("Mixer latency: %7.2f ms\n", (float)x*1e-3f);
            }

            x = aaxMixerGetSetup(cfg, AAX_MONO_EMITTERS);
            y = aaxMixerGetSetup(cfg, AAX_STEREO_EMITTERS);
            printf("Available mono emitters:   ");
            if (x == UINT_MAX) printf("infinite\n");
            else printf("%3i\n", x);
            printf("Available stereo emitters: ");
            if (y == UINT_MAX/2) printf("infinite\n");
            else printf("%3i\n", y);
            x = aaxMixerGetSetup(cfg, AAX_AUDIO_FRAMES);
            printf("Available audio-frames: ");
            if (x == UINT_MAX) printf("   infinite\n");
            else printf("%6i\n", x);

            printf("\nSupported Filters:\n ");
            for (i=1; i<AAX_FILTER_MAX; i++)
            {
                static int len = 1;
                if (aaxIsFilterSupported(cfg, _filter_s[i]))
                {
                    len += strlen(_filter_s[i])+1;   /* one for leading space */
                    if (len >= maximumWidth)
                    {
                        printf("\n ");
                        len = strlen(_filter_s[i])+1;
                    }
                    printf(" %s", _filter_s[i]);
                }
            }

            printf("\n\nSupported Effects:\n ");
            for (i=1; i<AAX_EFFECT_MAX; i++)
            {
                static int len = 1;
                if (aaxIsEffectSupported(cfg, _effect_s[i]))
                {
                    len += strlen(_effect_s[i])+1;   /* one for leading space */
                    if (len >= maximumWidth)
                    {
                        printf("\n ");
                        len = strlen(_effect_s[i])+1;
                    }
                    printf(" %s", _effect_s[i]);
                }
            }
            printf("\n\n");

            aaxDriverClose(cfg);
            aaxDriverDestroy(cfg);
        }
        else
        {
           enum aaxErrorType err = aaxGetErrorNo();
           printf("Error opening the default device: ");
           printf("%s\n\n", aaxGetErrorString(err));
        }
    }
    else {
        printf("\nDefault driver not available.\n\n");
    }

    return 0;
}
